src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,667 @@
+                                currOne.spi().engineDeriveKey(algorithm,
+                                        derivationSpec);
+                        // found a working KDFSpi
+                        this.theOne = currOne;
+                        return result;
+                    } catch (Exception e) {
+                        if (lastException == null) {
+                            lastException = e;
+                        }
+                        // try next one if available
+                        assert serviceIterator != null : "serviceIterator was null";
+                        currOne = getNext(serviceIterator, kdfParameters);
+                    }
+                }
+            } catch (InvalidAlgorithmParameterException e) {
+                throw e; // getNext reached end and have seen IAPE
+            } catch (NoSuchAlgorithmException e) {
+                // getNext reached end without finding an implementation
+                throw new InvalidAlgorithmParameterException(lastException);
+            }
+        }
+    }
+
+    private static Delegate getNext(Iterator<Service> serviceIter,
+                                    KDFParameters kdfParameters)
+            throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        // fetch next one if available
+        boolean hasOne = false;
+        while (serviceIter.hasNext()) {
+            Service s = serviceIter.next();
+            Provider prov = s.getProvider();
+            if (!JceSecurity.canUseProvider(prov)) {
+                // continue to next iteration
+                continue;
+            }
+            hasOne = true;
+            try {
+                Object obj = s.newInstance(kdfParameters);
+                if (!(obj instanceof KDFSpi)) {
+                    if (!skipDebug && pdebug != null) {
+                        pdebug.println(
+                                "obj was not an instance of KDFSpi (should not "
+                                + "happen)");
+                    }
+                    continue;
+                }
+                return new Delegate((KDFSpi) obj, prov);
+            } catch (NoSuchAlgorithmException nsae) {
+                // continue to the next provider
+                if (!skipDebug && pdebug != null) {
+                    pdebug.println(
+                            "The evaluated provider does not support the "
+                            + "specified algorithm and parameters");
+                }
+            }
+        }
+        if (!skipDebug && pdebug != null) {
+            pdebug.println(
+                    "No provider can be found that supports the "
+                    + "specified algorithm and parameters");
+        }
+        if (hasOne) throw new InvalidAlgorithmParameterException();
+        else throw new NoSuchAlgorithmException();
+    }
+
+    private static boolean checkSpiNonNull(Delegate d) {
+        return (d != null && d.spi() != null);
```

Comment:
Can `d.spi() == null` ever happen?

Reply:
If it's useless then remote it. We should be confidence in the logic. Adding unnecessary check will only confuse readers.